app/src/main/java/com/kamron/pogoiv/MainActivity.java

```diff
@@ -295,13 +309,42 @@ private void startPokeFly() {
         Intent intent = Pokefly.createIntent(this, trainerLevel, statusBarHeight, batterySaver);
         startService(intent);
 
-        pokeFlyRunning = true;
-
         if (settings.shouldLaunchPokemonGo()) {
             openPokemonGoApp();
         }
     }
 
+    private boolean isPokeFlyRunning() {
+        ActivityManager manager = (ActivityManager) getSystemService(Context.ACTIVITY_SERVICE);
+        for (ActivityManager.RunningServiceInfo serviceInfo : manager.getRunningServices(Integer.MAX_VALUE)) {
+            if (Pokefly.class.getName().equals(serviceInfo.service.getClassName())) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private void updateLaunchButtonText(boolean isPokeflyRunning) {
+        if (!hasAllPermissions()) {
+            launchButton.setText(R.string.main_permission);
+        } else if (isPokeflyRunning) {
+            launchButton.setText(R.string.main_stop);
+        } else {
+            launchButton.setText(R.string.main_start);
+        }
+    }
+
+    private boolean hasAllPermissions() {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && !Settings.canDrawOverlays(this)) {
+            return false;
+        }
+        if (ContextCompat.checkSelfPermission(MainActivity.this, Manifest.permission.WRITE_EXTERNAL_STORAGE)
+                == PackageManager.PERMISSION_DENIED) {
+            return false;
+        }
+        return true;
+    }
```

Comment:
Fine then, it's a matter of preference and expectations (especially regarding readability). ✅